src/config.rs

```diff
@@ -96,78 +96,179 @@ impl<T: Into<String>> From<T> for OverrideFile {
+                if file.toolchain.targets.is_some()
+                    || file.toolchain.components.is_some()
+                    || file.toolchain.profile.is_some()
+                {
                     bail!(
-                        "cannot specify both channel ({}) and path ({}) simultaneously",
-                        channel,
+                        "toolchain options are ignored for path toolchain ({})",
                         path.display()
                     )
                 }
-                (None, None) => None,
-            },
-            components: file.toolchain.components.unwrap_or_default(),
-            targets: file.toolchain.targets.unwrap_or_default(),
-            profile: file
-                .toolchain
-                .profile
-                .as_deref()
-                .map(dist::Profile::from_str)
-                .transpose()?,
+                // We -do not- support relative paths, they permit trivial
+                // completely arbitrary code execution in a directory.
+                // Longer term we'll not support path based toolchains at
+                // all, because they also permit arbitrary code execution,
+                // though with more challenges to exploit.
+                return Ok(Self::PathBased(PathBasedToolchainName::try_from(
+                    &path as &Path,
+                )?));
+            }
+            (Some(channel), Some(path)) => {
+                bail!(
+                    "cannot specify both channel ({}) and path ({}) simultaneously",
+                    channel,
+                    path.display()
+                )
+            }
+            (None, None) => cfg
+                .get_default()?
+                .ok_or(RustupError::ToolchainNotSelected)?,
+        };
+        Ok(match toolchain_name {
+            ToolchainName::Official(desc) => {
+                let components = file.toolchain.components.unwrap_or_default();
+                let targets = file.toolchain.targets.unwrap_or_default();
+                Self::Official {
+                    toolchain: desc,
+                    components,
+                    targets,
+                    profile: file
+                        .toolchain
+                        .profile
+                        .as_deref()
+                        .map(dist::Profile::from_str)
+                        .transpose()?,
+                }
+            }
+            ToolchainName::Custom(name) => {
+                if file.toolchain.targets.is_some()
+                    || file.toolchain.components.is_some()
+                    || file.toolchain.profile.is_some()
+                {
+                    bail!(
+                        "toolchain options are ignored for a custom toolchain ({})",
+                        name
+                    )
+                }
+                Self::Custom(name)
+            }
         })
     }
+
+    fn get_local_toolchain_name(self) -> LocalToolchainName {
```

Comment:
According to our [API guidelines](https://rust-lang.github.io/api-guidelines/naming.html), `get_*` should probably be replaced with `into_`.